OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java

```diff
@@ -2021,4 +1981,184 @@ private void avoidKeepForThroughMoving(List<RouteSegmentResult> result) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	private boolean hasSameTurnLanes(RouteSegmentResult prevSegm, RouteSegmentResult currentSegm) {
+		String turnLanesPrevSegm = getTurnLanesString(prevSegm);
+		String turnLanesCurrSegm = getTurnLanesString(currentSegm);
+		if (turnLanesPrevSegm == null || turnLanesCurrSegm == null) {
+			return false;
+		}
+		int[] uniqPrev = getUniqTurnTypes(turnLanesPrevSegm);
+		int[] uniqCurr = getUniqTurnTypes(turnLanesCurrSegm);
+		if (uniqPrev.length != uniqCurr.length) {
+			return false;
+		}
+		for (int i = 0; i < uniqCurr.length; i++) {
+			if (uniqPrev[i] != uniqCurr[i]) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	private boolean hasAllowedLanes(int mainTurnType, int[] lanesArray, int startActiveIndex, int endActiveIndex) {
+		if (lanesArray.length == 0 || startActiveIndex > endActiveIndex) {
+			return false;
+		}
+		int[] activeLines = new int[endActiveIndex - startActiveIndex + 1];
+		for (int i = startActiveIndex, j = 0; i <= endActiveIndex; i++, j++) {
+			activeLines[j] = lanesArray[i];
+		}
+		boolean left = TurnType.isLeftTurnNoUTurn(mainTurnType);
+		boolean right = TurnType.isRightTurnNoUTurn(mainTurnType);
+		Set<Integer> leftDirections = new HashSet<Integer>(){{
+			add(TurnType.TL);
+			add(TurnType.TSHL);
+			add(TurnType.TSLL);
+		}};
+		Set<Integer> rightDirections = new HashSet<Integer>(){{
+			add(TurnType.TR);
+			add(TurnType.TSHR);
+			add(TurnType.TSLR);
+		}};
+		Set<Integer> sharpRightDirections = new HashSet<Integer>(){{
+			add(TurnType.TSHR);
+			add(TurnType.TRU);
+		}};
+		Set<Integer> sharpLeftDirections = new HashSet<Integer>(){{
+			add(TurnType.TSHL);
+			add(TurnType.TU);
+		}};
+		for (int i = 0; i < activeLines.length; i++) {
+			int turnType = TurnType.getPrimaryTurn(activeLines[i]);
+			if (turnType == mainTurnType) {
+				return true;
+			}
+			if (left && leftDirections.contains(turnType)) {
+				return true;
+			}
+			if (right && rightDirections.contains(turnType)) {
+				return true;
+			}
+			if (mainTurnType == TurnType.C && TurnType.isSlightTurn(turnType)) {
+				return true;
+			}
+			if (sharpLeftDirections.contains(mainTurnType) && sharpLeftDirections.contains(turnType)) {
+				return true;
+			}
+			if (sharpRightDirections.contains(mainTurnType) && sharpLeftDirections.contains(turnType)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	private TurnType getActiveTurnType(int[] lanes, boolean leftSide, TurnType oldTurnType) {
+		if (lanes == null || lanes.length == 0) {
+			return oldTurnType;
+		}
+		int tp = oldTurnType.getValue();
+		int cnt = 0;
+		for (int k = 0; k < lanes.length; k++) {
+			int ln = lanes[k];
+			if ((ln & 1) > 0) {
+				int[] oneActiveLane = {lanes[k]};
+				if (hasAllowedLanes(oldTurnType.getValue(), oneActiveLane, 0, 0)) {
+					tp = TurnType.getPrimaryTurn(lanes[k]);
+				}
+				cnt++;
+			}
+		}
+		TurnType t = TurnType.valueOf(tp, leftSide);
+		if (cnt == 3 && TurnType.isSlightTurn(t.getValue())) {
```

Comment:
here we need to add comment what 3 means

Reply:
Old code, moved from 
https://github.com/osmandapp/OsmAnd/blob/7d3994d0844a28b94e85aa3de28ff91ce984f94f/OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java#LL1446C17-L1446C17